sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala

```diff
@@ -234,7 +234,22 @@ private[sql] class DefaultWriterContainer(
     executorSideSetup(taskContext)
     val configuration = SparkHadoopUtil.get.getConfigurationFromJobContext(taskAttemptContext)
     configuration.set("spark.sql.sources.output.path", outputPath)
-    val writer = outputWriterFactory.newInstance(getWorkPath, dataSchema, taskAttemptContext)
+    val writer = try {
+      outputWriterFactory.newInstance(getWorkPath, dataSchema, taskAttemptContext)
+    } catch {
+      case e: org.apache.hadoop.fs.FileAlreadyExistsException =>
+        if (outputCommitter.isInstanceOf[parquet.DirectParquetOutputCommitter]) {
+          // Spark-11382: DirectParquetOutputCommitter is not idempotent, meaning on retry
+          // attempts, the task will fail because the output file is created from a prior attempt.
+          // This often means the most visible error to the user is misleading. Augment the error
+          // to tell the user to look for the actual error.
+          throw new SparkException("The output file already exists but this could be due to a " +
+            "failure from a earlier attempt. Look through the earlier logs for the first " +
+            "error.\n  File exists error: " + e)
```

Comment:
I explicitly did not do that because that makes it really seem like the root cause is "file already exists". This error would be in the middle of a big stack trace. We can go either way though.